### ISSUE TITLE
weird criterion to decide if needed to adjust the padding size

### ISSUE THREAD

**Issue by hyusterr** [huggingface/transformers#35599]
https://github.com/huggingface/transformers/blob/241c04d36867259cdf11dbb4e9d9a60f9cb65ebc/src/transformers/trainer_pt_utils.py#L93

When working with `Trainer` and model output with `TokenClassiferOutput`, the `Trainer` collect outputs batch by batch in `evaluate_loop` with `nested_concat`.
The criterion to decide whether it is needed to adjust the shape is by tensor.shape[1]. 
However, if user does not exclude `attentions` with `ignore_key` beforehand, the shape[1] of the attentions tensor is always the same. Leading to error occurred with `torch.cat`. 
I think this design is kind of weird, especially because this behavior is not emphasized in the token classification tutorial: https://huggingface.co/docs/transformers/tasks/token_classification
